docs/proposals/dehydration.md

```diff
@@ -0,0 +1,147 @@
+# Cluster Dehydration
+
+**Author**: Marvin Beckers (@embik)
+
+**Status**: Draft proposal.
+
+## Goals
+
+Dehydration is a functionality for KKP user clusters that will allow to temporarily "suspend" a cluster to free up cloud resources used by the cluster. The feature will allow to optimize resource consumption and subsequent spending on cloud resources. State (e.g. storage) should not be affected by this functionality.
+
+## Non-Goals
+
+Migrating dehydrated user clusters between seeds is out of scope for this proposal. User clusters can only be re-started on the same seed.
```

Comment:
Have you maybe considered adding similar functionality at the MachineDeployment level? I can image some use cases in scenarios with taints or node selectors.

Another thing that came to my mind was introducing schedules for hydration. In short, letting the user define hours (potentially even days) in which cluster is hydrated. But I assume it's to early to talk about such solution. I just wanted to mention that as a potential extension that could be build on top of what this proposal describes.

Reply:
> Have you maybe considered adding similar functionality at the MachineDeployment level? I can image some use cases in scenarios with taints or node selectors.

I'm afraid I don't know what functionality you are referring to here.

> Another thing that came to my mind was introducing schedules for hydration. In short, letting the user define hours (potentially even days) in which cluster is hydrated. But I assume it's to early to talk about such solution. I just wanted to mention that as a potential extension that could be build on top of what this proposal describes.

I agree with this and actually thought about it as well, but forgot to add it to the proposal. I can however see us adding a `DehydrationSchedule` resource that could even support label matching, so you can dehydrate all of our `env=staging` labelled clusters during off-hours. WDYT?

Reply:
> I'm afraid I don't know what functionality you are referring to here.

I was thinking of de-hydrating only specific pool of nodes in the cluster.

Reply:
> > I'm afraid I don't know what functionality you are referring to here.
> 
> I was thinking of de-hydrating only specific pool of nodes in the cluster.

I'm not sure that makes a lot of sense, because pods will attempt to restart on other nodes then, and it's very simple to replicate this by just scaling down the specific pool. Some pods might be rescheduled, some might not be, but it's not going to be stable behaviour. But I'm open to this if others would find it useful.

Reply:
> 

You are right. It could cause a lot of trouble without proper taints/tolerations in place and users can achieve the same result by scaling down the pool.

Reply:
> > > I'm afraid I don't know what functionality you are referring to here.
> > 
> > 
> > I was thinking of de-hydrating only specific pool of nodes in the cluster.
> 
> I'm not sure that makes a lot of sense, because pods will attempt to restart on other nodes then, and it's very simple to replicate this by just scaling down the specific pool. Some pods might be rescheduled, some might not be, but it's not going to be stable behaviour. But I'm open to this if others would find it useful.

You are right. It could cause a lot of trouble without proper taints/tolerations in place and users can achieve the same result by scaling down the pool.

Reply:
> I agree with this and actually thought about it as well, but forgot to add it to the proposal. I can however see us adding a DehydrationSchedule resource that could even support label matching, so you can dehydrate all of our env=staging labelled clusters during off-hours. WDYT?

Sounds good to me.